wmanley/stb-tester-service
# Cache results in MongoDB

To allow querying the results faster to allow looking up results by job for instance.

**TODO**:
- [x] Write `test_that_generation_id_is_generated_if_not_specified`
- [x] Write `test_that_generation_id_is_persisted_between_invocations`
- [x] Add tests for parsing results directories
- [x] Add tests around mongodb robustness WRT unclean shutdown and restart
- [x] Parse results that are already there in the filesystem but are not in mongo (to cope with unclean shutdown and results generated prior to these changes).
- [x] Factor other UI changes into another pull-request
- [x] Write tests for `AtomicTree.watch`
- [x] Tests: Don't duplicate `find_unused_port_racy`
- [x] Tests: Don't duplicate `MongoDB` test helper
- [x] Fix stb-tester stb-tester/stb-tester#260 to report `test_case`, not `test` in `StateWatch`

**And then in future PRs but not this one**:
- Import results in background, preventing tests from running during this time.
- Merge stb-tester/stb-tester#260
- Make `watch` return a `Watch` object which we can use to unregister a watch and get the initial value.  It should be a `contextmanager`:

        q = Queue()
        def callback(old, new):
            q.push(new)
        with tree.watch(("state", "line_no"), callback) as initial_value:
            send_to_web_ui(inital_value)
            while True:            
                send_to_web_ui(q.pop(block=True))

- Rewrite results view in angular
- Write results into job specific directories
- Add colours to the job progress bar (#20)
- Add endpoint for retrieving results for a given job (particularly the current one)
- Turn off `stbt batch`'s html generating (it's slow in large directories)
